js/RichText/UserTemplatesParameters.js

```diff
@@ -0,0 +1,209 @@
+            );
+            resolve(items);
+         }
+      );
+   }
+
+   /**
+    * Recursive function to parse available parametes into a format that can
+    * be handled by the autocompletion
+    *
+    * @private
+    *
+    * @param {Array} parameters
+    * @param {string} prefix
+    *
+    * @returns {Array} Parsed parameters
+    */
+   parseParameters(parameters, prefix = "") {
+      const parsed_parameters = [];
+      const that = this;
+
+      parameters.forEach(parameter => {
+         // Add prefix, needed when we go down recursivly so we don't lose track
+         // of the main item (e.g ticket.entity.name instead of entity.name)
+         if (prefix.length > 0) {
+            parameter.key = prefix + "." + parameter.key;
+         }
+
+         switch (parameter.type) {
+            // Add a simple attribute to autocomplete
+            case 'AttributeParameter': {
+               let value = '{{ ' + parameter.key;
+               if (parameter.filter && parameter.filter.length) {
+                  value += ' | ' + parameter.filter;
+               }
+               value += " }}";
+
+               parsed_parameters.push({
+                  type: 'autocompleteitem',
+                  value: value,
+                  text: value + ' - ' + parameter.label,
+               });
+               break;
+            }
+
+            // Recursivly parse parameters of the given object
+            case 'ObjectParameter': {
+               parsed_parameters.push(...that.parseParameters(parameter.properties, parameter.key));
+               break;
+            }
+
+            // Add a possible loop to the autocomplete, with extra autocomplete
+            // support for the content of the array.
+            case 'ArrayParameter': {
+               parsed_parameters.push({
+                  type: 'autocompleteitem',
+                  value: '{% for ' + parameter.items_key + ' in ' + parameter.key + ' %}',
+                  text: '{% for ' + parameter.items_key + ' in ' + parameter.key + ' %} - ' + parameter.label,
+               });
+
+               // Push content of array, hidden by default unless the parent loop exist in the editor
+               const content = that.parseParameters([parameter.content]);
+               parsed_parameters.push(
+                  ...content.map(
+                     function(item) {
+                        item.show = '{% for ' + parameter.items_key + ' in ' + parameter.key + ' %}';
+                        return item;
+                     }
+                  )
+               );
+               break;
+            }
+         }
+      });
+
+      return parsed_parameters;
+   }
+
+   /**
+    * Add mention to selected user in editor.
+    *
+    * @private
+    *
+    * @param {AutocompleterInstanceApi} autocompleteApi
+    * @param {Range} range
+    * @param {string} value
+    *
+    * @returns {void}
+    */
+   insertTwigContent(autocompleteApi, range, value) {
```

Comment:
Here you should encapsulate twig tags into `<span contenteditable="false"></span>`. That would permit to protect them from alteration.

Reply:
Seemed a good idea at first but this will prevent the users from editing twig filters (changing date format, setting to uppercase, ...)